Update to latest eslint v8

gregaubert opened the pull request:
I've updated to all the latest versions and tried it on SC.

On SC we found around 70 new issues that were actually relevant, mostly raising some useless props that were not used anymore and easy to clean and some errors related to no-conditions-in-test that I cleaned.

There is also one new react rule ([react/jsx-no-leaked-render](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-no-leaked-render.md)) that seemed valuable but that I disabled because it generated a huge amount of false positives. You can find the PR with the SC changes here: https://github.com/SonarSource/sonarcloud-core/pull/3655

Regarding the warnings, we had the same amount as before.

Could you guys have a try on SQ and tell me if you think it makes sense to release like this or if we should disable or change some rules to warn? You can use version 2.0.0-rc to try it.

gregaubert committed:
Add npmrc file to ease publication

gregaubert committed:
Node 12.22.0 as minimum supported version

gregaubert committed:
Update eslint to v8

gregaubert committed:
Update plugins that don't need any rules update

gregaubert committed:
Update React plugin

- Add react/hook-use-state rule
- Add react/no-arrow-function-lifecycle rule
- Add react/no-namespace rule
- allowExpressions in react/jsx-no-useless-fragment rule
- Disable react/jsx-no-leaked-render rule because of the many FP

gregaubert committed:
Update jest plugin

- Add jest/prefer-equality-matcher rule
- Add jest/prefer-comparison-matcher rule
- Replace deprecated jest/no-if rule by jest/no-conditional-in-test

gregaubert committed:
v2.0.0-rc

gregaubert commented:
@jeremy-davis-sonarsource and @mathieu-suen-sonarsource I'm also thinking about releasing another version that would add eslint-plugin-testing-library and eslint-plugin-jest-dom with the recommended profiles. Bot those plugins are for React testing library, we use them in SC.
Are you also using them in SQ? Do you think it would make sense to have them by default?

jeremy-davis-sonarsource commented:
> @jeremy-davis-sonarsource and @mathieu-suen-sonarsource I'm also thinking about releasing another version that would add eslint-plugin-testing-library and eslint-plugin-jest-dom with the recommended profiles. Bot those plugins are for React testing library, we use them in SC. Are you also using them in SQ? Do you think it would make sense to have them by default?

We use RTL as well, so those would make sense.

As for the new eslint, I ran it locally, and the only issues I get are `no-conditional-in-test` and `no-conditional-expect` (plus the odd `no-loss-of-precision` in measures-test)... I think it makes sense to update, but at the same time, I'm not convinced by those two rules. I'd lower them to warnings.

gregaubert commented:
For the `no-conditional-in-test`, we had a few places where it was raised because we had conditional in some mocks, those are almost FP I would say...
And a few others because we were checking that something was defined before accessing a property, like: `expect(something && something.that).toBe(3)` and I fixed those by doing `expect(something!.that).toBe(3)`, because in the test you know it's defined...
But I agree that the rule is a bit noisy because of the mocks... I'll move it to warning.

Regarding `no-conditional-expect` we didn't get any of those, but it seems it can be a valuable rule from the example they give on the description, really avoiding having test that pass when there is actually no assertion done... Are you sure we should move it to warn? On what kind of stuff did it trigger?

For the `no-loss-of-precision`, it's was actually on unused code that I just dropped, maybe it's the same on SQ?

jeremy-davis-sonarsource commented:
Seems most of the `no-conditional-expect`s are for `it.each` tests, but they can be refactored.

don't worry about `no-loss-of-precision`.

=> I'm fine if we only downgrade `no-conditional-in-test` to warning. :metal:

gregaubert commented:
It's not great if it's rising errors on `it.each`, is it legit errors?

Ok so for now I'm just downgrading `no-condition-in-test`, will release like that and then do a new version with RTL plugins!
Thanks for the review!

gregaubert committed:
Change no-conditional-in-test from error to warn

gregaubert committed:
v2.0.0

gregaubert commented:
@jeremy-davis-sonarsource @mathieu-suen-sonarsource Can I proceed with the merge and the release ?

jeremy-davis-sonarsource commented:
> It's not great if it's rising errors on `it.each`, is it legit errors?

Yes, we have conditions in the test instead of expected values in the dataset, which isn't a great pattern
e.g.:
```
it.each([...])(() => {
  if (value == something) expect().toBeUndefined();
  else expect().toBe(toto)
})
```
instead of 
```
it.each([[..., undefined], [..., toto]])
```
